File: ruby/palindrome-products/palindrome_products.rb
require "set"

Palindrome = Struct.new(:value, :factors)

class Palindromes
  def self.palindrome?(n)
    str = n.to_s
    return str == str.reverse
  end

  def initialize(min_factor: 1, max_factor: 9)
    @min_factor = min_factor
    @max_factor = max_factor

    @values_to_factors = Hash.new { |h, k| h[k] = Set.new }
  end

  def generate
    (@min_factor..@max_factor).to_a.repeated_combination(2) do |i, j|
      value = i * j

      if Palindromes.palindrome?(value)
        @values_to_factors[value] << [i, j].sort
      end
    end
  end

  def get(n)
    Palindrome.new(n, @values_to_factors[n].to_a)
  end

  def largest
    get(@values_to_factors.keys.max)
  end

  def smallest
    get(@values_to_factors.keys.min)
  end
end

File: javascript/palindrome-products/palindrome-products.js
export class Palindromes {
    static generate(kwargs) {
        return new Palindromes(kwargs);
    }

    static isPalindrome(n) {
        let chars = [...n.toString()];
        let reverse = chars.slice().reverse();

        return chars.join("") == reverse.join("");
    }

    constructor(kwargs) {
        this.valuesToFactors = new Map;

        let minFactor = kwargs.minFactor || 1;
        let maxFactor = kwargs.maxFactor || 9;

        if (minFactor > maxFactor)
            throw "min must be <= max";

        let range = Array.range(minFactor, maxFactor);

        for (const [i, j] of range.eachRepeatedCombination(2)) {
            let value = i * j;

            if (Palindromes.isPalindrome(value)) {
                let factors = [i, j].sort();

                if (this.valuesToFactors.has(value))
                    this.valuesToFactors.get(value).add(factors);
                else
                    this.valuesToFactors.set(value, new Set([factors]));
            }
        }
    }

    get(n) {
        if (this.valuesToFactors.has(n))
            return {
                value: n,
                factors: [...this.valuesToFactors.get(n)]
            };
        else
            return {value: null, factors: []};
    }

    getWithFunc(func) {
        let value = func.call(null, ...this.valuesToFactors.keys());
        return this.get(value);
    }

    get largest() {
        return this.getWithFunc(Math.max);
    }

    get smallest() {
        return this.getWithFunc(Math.min);
    }
}

// Source:
// <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from>
Array.range = function(start, stop, step = 1) {
    return Array.from(
        {length: (stop - start) / step + 1},
        (_, i) => start + i * step
    );
}

Array.prototype.eachRepeatedCombination = function*(k) {
    if (k < 0)
        throw "Size must be positive";

    if (k == 0) {
        yield [];
    } else if (k == 1) {
        for (const x of this)
            yield [x];
    } else {
        for (let i = 0; i < this.length; ++i)
            for (const c of this.slice(i).eachRepeatedCombination(k - 1))
                yield [this[i], ...c];
    }
}
